README.md
# gt-gin-vue

POST   /auth/login      登录  
curl http://host:port/auth/login -d '{"name":"gt","pwd":"123"}'

POST   /auth/register   注册  
curl http://host:port/auth/register -d '{"username":"gt","pwd":"123"}'

GET    /user/test       测试token  
curl http://host:port/user/test -H "token:abc"

GET    /user/resetpwd   重置用户密码  
curl http://host:port/user/resetpwd -H "token:abc"(此处token内会存储用户id)

POST   /user/delete     删除用户  
curl http://host:port/user/delete -H "token:abc" -d "name=gt"

POST   /user/update     更新用户信息  
curl http://host:port/user/test -H "token:abc" -d "name=gt&pwd=123" 



changelog.md
2020-05-22
=========
```

实现用户增删改查，token鉴权

```

2020-05-20
=========
```

middleware 中间件
api 给前端调用的的api接口
model 业务逻辑+数据类型

```

test/test.go
package main

import (

	_ "github.com/go-sql-driver/mysql"
	"github.com/jinzhu/gorm"
	"fmt"
)

type User struct {
	Id           uint `gorm:"primary_key"`
	Username     string
	Password     string
}

func main() {

	db, err := gorm.Open("mysql", "gt:abc123@tcp(127.0.0.1:3306)/gt_gin_vue?charset=utf8")
	if err != nil {
		fmt.Println("err")
		return
	}
	defer db.Close()
	db.CreateTable(&User{})
	user := User {
		Id:1,
		Username:"usme",
		Password:"pwd",
	}
	fmt.Println(user);
	if err := db.Create(&user).Error; err != nil {
		fmt.Println("插入失败", err)
	}
}


api/api.go
package api

import (
	myjwt "gt-gin-vue/middleware/jwt"
	"gt-gin-vue/model"
	"log"
	"net/http"
	"time"

	jwtgo "github.com/dgrijalva/jwt-go"
	"github.com/gin-gonic/gin"
)

// 注册信息
type RegistInfo struct {
	UserName string `json:"username"`
	Pwd string `json:"pwd"`
}

// Register 注册用户
func RegisterUser(c *gin.Context) {
	var registerInfo RegistInfo
	if c.BindJSON(&registerInfo) == nil {
		err := model.Register(registerInfo.UserName, registerInfo.Pwd)
		if err == nil {
			c.JSON(http.StatusOK, gin.H{
				"status": 0,
				"msg":    "注册成功！",
			})
		} else {
			c.JSON(http.StatusOK, gin.H{
				"status": -1,
				"msg":    "注册失败:" + err.Error(),
			})
		}
	} else {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "解析数据失败！",
		})
	}
}

// LoginResult 登录结果结构
type LoginResult struct {
	Token string `json:"token"`
	model.User
}

// Login 登录
func Login(c *gin.Context) {
	var loginReq model.LoginReq
	if c.BindJSON(&loginReq) == nil {
		isPass, user, err := model.LoginCheck(loginReq)
		if isPass {
			generateToken(c, user)
		} else {
			c.JSON(http.StatusOK, gin.H{
				"status": -1,
				"msg":    "验证失败," + err.Error(),
			})
		}
	} else {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "json 解析失败",
		})
	}
}

// 生成令牌
func generateToken(c *gin.Context, user model.User) {
	j := &myjwt.JWT{
		[]byte("GaoTaoLearn"),
	}
	claims := myjwt.CustomClaims{
		user.Id,
		user.Username,
		jwtgo.StandardClaims{
			NotBefore: int64(time.Now().Unix() - 1000), // 签名生效时间
			ExpiresAt: int64(time.Now().Unix() + 3600), // 过期时间 一小时
			Issuer:    "GaoTaoLearn",                   //签名的发行者
		},
	}

	token, err := j.CreateToken(claims)

	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    err.Error(),
		})
		return
	}

	log.Println(token)

	data := LoginResult{
		User:  user,
		Token: token,
	}
	c.JSON(http.StatusOK, gin.H{
		"status": 0,
		"msg":    "登录成功！",
		"data":   data,
	})
	return
}

//下面的函数若要有权限问题需要再加中间件

func TestToken(c *gin.Context) {
	//若要加admin权限，需要解析claims
	claims := c.MustGet("claims").(*myjwt.CustomClaims)
	if claims != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": 0,
			"msg":    "token有效",
			"data":   claims,
		})
	}
}

func DeleteUser(c *gin.Context) {
	//若要加admin权限，需要解析claims
	claims := c.MustGet("claims").(*myjwt.CustomClaims)

	name := c.PostForm("name")
	if name == "" {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "未传递用户名",
		})
		return
	}
	err := model.Delete(name, claims.Id)

	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "删除用户失败 " + err.Error(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"status": 0,
		"msg":    "成功删除用户 " + name,
	})
}

func InfoUsers(c *gin.Context) {
	//若要加admin权限，需要解析claims
	//claims := c.MustGet("claims").(*myjwt.CustomClaims)
}

func UpdateUser(c *gin.Context) {
	//若要加admin权限，需要解析claims
	claims := c.MustGet("claims").(*myjwt.CustomClaims)
	name := c.PostForm("name")
	pwd  := c.PostForm("pwd")
	if name == "" && pwd == "" {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "无任何修改参数",
		})
	} 
	user := model.User {
		Id:claims.Id,
		Password: pwd,
		Username: name,
	}
	retUser, err := model.Update(user)

	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "更新用户数据失败 " + err.Error(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"status": 0,
		"msg":    "更新用户数据成功 ",
		"data": retUser,
	})
}

func ResetPassword(c *gin.Context) {
	//若要加admin权限，需要解析claims
	claims := c.MustGet("claims").(*myjwt.CustomClaims)

	err := model.ResetPwd(claims.Id)

	if err != nil {
		c.JSON(http.StatusOK, gin.H{
			"status": -1,
			"msg":    "重置密码失败 " + err.Error(),
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"status": 0,
		"msg":    "重置密码为123",
	})
}


middleware/jwt/jwt.go
package jwt

import (
	"errors"
	"log"
	"net/http"
	"time"

	"github.com/dgrijalva/jwt-go"
	"github.com/gin-gonic/gin"
)

// JWTAuth 中间件，检查token
func JWTAuth() gin.HandlerFunc {
	return func(c *gin.Context) {
		// curl http://219.83.160.146:8080/data/dataByTime -H "token:..."
		token := c.Request.Header.Get("token")
		if token == "" {
			c.JSON(http.StatusOK, gin.H{
				"status": -1,
				"msg":    "请求未携带token，无权限访问",
			})
			c.Abort()
			return
		}

		log.Print("get token: ", token)

		j := NewJWT()
		// parseToken 解析token包含的信息
		claims, err := j.ParseToken(token)
		if err != nil {
			if err == TokenExpired {
				c.JSON(http.StatusOK, gin.H{
					"status": -1,
					"msg":    "授权已过期",
				})
				c.Abort()
				return
			}
			c.JSON(http.StatusOK, gin.H{
				"status": -1,
				"msg":    err.Error(),
			})
			c.Abort()
			return
		}
		// 继续交由下一个路由处理,并将解析出的信息传递下去
		c.Set("claims", claims)
	}
}

// JWT 签名结构
type JWT struct {
	SigningKey []byte
}

// 一些常量
var (
	TokenExpired     error  = errors.New("Token is expired")
	TokenNotValidYet error  = errors.New("Token not active yet")
	TokenMalformed   error  = errors.New("That's not even a token")
	TokenInvalid     error  = errors.New("Couldn't handle this token:")
	SignKey          string = "GaoTaoLearn"
)

// 载荷，可以加一些自己需要的信息
type CustomClaims struct {
	Id    uint `json:"userid"`
	Username  string `json:"username"`
	jwt.StandardClaims
}

// 新建一个jwt实例
func NewJWT() *JWT {
	return &JWT{
		[]byte(GetSignKey()),
	}
}

// 获取signKey
func GetSignKey() string {
	return SignKey
}

// 这是SignKey
func SetSignKey(key string) string {
	SignKey = key
	return SignKey
}

// CreateToken 生成一个token
func (j *JWT) CreateToken(claims CustomClaims) (string, error) {
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(j.SigningKey)
}

// 解析Tokne
func (j *JWT) ParseToken(tokenString string) (*CustomClaims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &CustomClaims{}, func(token *jwt.Token) (interface{}, error) {
		return j.SigningKey, nil
	})
	if err != nil {
		if ve, ok := err.(*jwt.ValidationError); ok {
			if ve.Errors&jwt.ValidationErrorMalformed != 0 {
				return nil, TokenMalformed
			} else if ve.Errors&jwt.ValidationErrorExpired != 0 {
				// Token is expired
				return nil, TokenExpired
			} else if ve.Errors&jwt.ValidationErrorNotValidYet != 0 {
				return nil, TokenNotValidYet
			} else {
				return nil, TokenInvalid
			}
		}
	}
	if claims, ok := token.Claims.(*CustomClaims); ok && token.Valid {
		return claims, nil
	}
	return nil, TokenInvalid
}

// 更新token
func (j *JWT) RefreshToken(tokenString string) (string, error) {
	jwt.TimeFunc = func() time.Time {
		return time.Unix(0, 0)
	}
	token, err := jwt.ParseWithClaims(tokenString, &CustomClaims{}, func(token *jwt.Token) (interface{}, error) {
		return j.SigningKey, nil
	})
	if err != nil {
		return "", err
	}
	if claims, ok := token.Claims.(*CustomClaims); ok && token.Valid {
		jwt.TimeFunc = time.Now
		claims.StandardClaims.ExpiresAt = time.Now().Add(1 * time.Hour).Unix()
		return j.CreateToken(*claims)
	}
	return "", TokenInvalid
}

model/user.go
package model

import (
	_ "github.com/go-sql-driver/mysql"
	"github.com/jinzhu/gorm"

	"errors"
	"log"
)

var (
	db *gorm.DB
	db_err error
)

// User 用户类 在数据库中表名默认会加上s，所有的字段名都会自动匹配起来，只是大写都变成了小写
// 可以自己在后面改名称用`gorm:"column:username"`如果列名称已经定了就不能插入了
type User struct {
	Id 	     uint `gorm:"column:id; primary_key; AUTO_INCREMENT"`
	Username string `gorm:"column:username"` 
	Password string `gorm:"column:password"`
}

// LoginReq 登录请求参数类
type LoginReq struct {
	Name string `json:"name"`
	Pwd   string `json:"pwd"`
}

func OpenMysql() (*gorm.DB, error) {
	db, db_err = gorm.Open("mysql", "gt:123@(127.0.0.1:3306)/dbtest?charset=utf8&parseTime=True&loc=Local")
	if db_err != nil {
		return nil, db_err
	}
	db.LogMode(true)//开启sql debug 模式

	ret := db.HasTable(&User{})
	if ret == false {
		db.CreateTable(&User{})//初始化一个空的表
	}
	return db, nil
}


func Add(user *User) error {
	
	if err := db.Create(user).Error; err != nil {
		log.Fatal("插入失败:", err)
		return err
	}

	return nil
}

// Register 插入用户，先检查是否存在用户，如果没有则存入
func Register(username string, pwd string) error {
	if CheckUser(username) {
		return errors.New("用户已存在！")
	}
	user := &User {
		Username:username,
		Password:pwd,
	}
	return Add(user);
}

// CheckUser 检查用户是否存在
func CheckUser(username string) bool {

	ret := db.Where("username = ?", username).First(&User{}).RecordNotFound()
	if ret == true {
		return false
	}
	return true
}

func Delete(username string, id uint) error {

	if !CheckUser(username) {
		return errors.New("该用户不存在")
	}
	// 删除表中所有的数据
	// u := User {}
	// db.Delete(&u)
	u := User {
		Username:username, 
		Id:id,
	}
	err := db.Delete(&u).Error
	return err
}

func Update(updateUser User) (User, error) {

	//根据前面的user的主键不变，修改成后面user里面的字段
	err := db.Model(&updateUser).Updates(&updateUser).Error
	return updateUser, err
}

// LoginCheck 登录验证
func LoginCheck(loginReq LoginReq) (bool, User, error) {
	user := User {}
	ret := db.Where("username = ?", loginReq.Name).First(&user).RecordNotFound()
	if ret == true {
		return false, user, errors.New("用户名未注册")
	}
	if user.Password != loginReq.Pwd {
		return false, user, errors.New("密码错误")
	}
	return true, user, nil
}

//ResetPwd 重置密码
func ResetPwd(id uint) error {

	user := User {Id:id}
	err := db.Model(&user).Update("password", "123").Error
	return err
}
